kernel/src/process/signal/pauser.rs

```diff
@@ -259,7 +255,185 @@ impl SigQueueObserver {
 impl Observer<SigEvents> for SigQueueObserver {
     fn on_events(&self, _: &SigEvents) {
         self.set_interrupted();
-        self.pauser.wait_queue.wake_all();
+        self.pauser_queue.wait_queue.wake_all();
+    }
+}
+
+struct SigMaskGuard<'a> {
+    old_mask: SigMask,
+    thread: &'a PosixThread,
+}
+
+impl<'a> SigMaskGuard<'a> {
+    fn new(thread: &'a PosixThread, waiter_mask: SigMask) -> Self {
+        let sig_mask = thread.sig_mask();
+        let old_mask = sig_mask.load(Ordering::Relaxed);
+        let new_mask = old_mask + waiter_mask;
+        sig_mask.store(new_mask, Ordering::Relaxed);
+        Self { old_mask, thread }
+    }
+}
+
+impl<'a> Drop for SigMaskGuard<'a> {
+    fn drop(&mut self) {
+        self.thread
+            .sig_mask()
+            .store(self.old_mask, Ordering::Relaxed)
+    }
+}
+
+/// A `Pauser` allows pausing the execution of the current thread until certain conditions are reached.
+///
+/// A `Pauser` will always bind the thread that creates it, and work with the thread-directed [`Waker`] in
+/// the thread. `Pauser` is implemented with [`Waiter`]. But unlike its [`Waiter`] relatives,
+/// `Pauser` is aware of POSIX signals:
+/// if a thread paused by a `Pauser` receives a signal, then the thread will resume its execution.
+///
+///  A `Pauser` cannot resume its binding thread by itself. Other threads should use the [`Waker`]
+/// in its binding thread to wake the thread up.
+///
+/// [`Waker`]: ostd::sync::Waker
+///
+/// # Example
+///
+/// Here is how the current thread can be put to sleep with a `Pauser`.
+///
+/// ```no_run
+/// let pauser = Pauser::new().with_mask(SigMask::new_full());
+/// // Pause the execution of the current thread until a user-given condition is met
+/// // or the current thread is interrupted by a signal.
+/// let res = pauser.pause_until(|| {
+///     if cond() {
+///         Some(())
+///     } else {
+///         None
+///     }
+/// });
+/// match res {
+///     Ok(_) => {
+///         // The user-given condition is met...
+///     }
+///     Err(EINTR) => {
+///         // A signal is received...
+///     }
+///     _ => unreachable!()
+/// }
+/// ```
+///
+/// Let's assume that another thread has access to the thread that be paused by this `Pauser`.
+/// Then, this second thread can resume the execution of the first thread
+/// even when `cond()` does not return `true`.
+///
+/// ```no_run
+/// let waker = paused_thread.waker();
+/// waker.wake_up();
+/// ```
+pub struct Pauser {
+    current_thread: Arc<Thread>,
+    waiter: Waiter,
+    sig_mask: SigMask,
```

Comment:
I'm wondering why we need to specify the mask for the pauser and the pauser queue. It looks like the only valid use case is to implement the `rt_sigsuspend' system call. Why don't we do the following:
```Rust
pub fn with_sigmask<R>(mask: SigMask, f: impl FnOnce() -> R) -> R;

// In `rt_sigsuspend':
with_sigmask(mask, || pauser.pause());
```

This is logically equivalent to making the `SigMaskGuard` public, but I use the closure pattern because if we create two `SigMaskGuard` and drop them in reverse order, the mask will get messed up.

I think this way the pauser/pause queue implementation can be cleaned up a lot, allowing for more code reuse. We can also have something like `wait_until_or_signalled` directly on the `Waiter`.

The downside might be that we have to pay twice for the cost of getting the current thread (once in `with_sigmask` and once in `pauser.pause`). But considering the really few use cases of the mask on the pauser, I guess this is acceptable.